Hmm, the Playstation build failure definitely looks related:
```
C:/BW/PlayStation-Build-EWS/build/WebKitBuild/Release/WTF/Headers\wtf/Ref.h:63:16: error: member access into incomplete type 'WebCore::StringCallback'
   63 |             ptr->deref();
      |                ^
C:/BW/PlayStation-Build-EWS/build/WebKitBuild/Release/WTF/Headers\wtf/RefPtr.h:61:47: note: in instantiation of member function 'WTF::DefaultRefDerefTraits<WebCore::StringCallback>::derefIfNotNull' requested here
   61 |     ALWAYS_INLINE ~RefPtr() { RefDerefTraits::derefIfNotNull(PtrTraits::exchange(m_ptr, nullptr)); }
      |                                               ^
C:/BW/PlayStation-Build-EWS/build/Source/WebCore/page\PageConsoleClient.h:53:22: note: in instantiation of member function 'WTF::RefPtr<WebCore::StringCallback>::~RefPtr' requested here
   53 | class WEBCORE_EXPORT PageConsoleClient final : public JSC::ConsoleClient, public CanMakeCheckedPtr<PageConsoleClient> {
      |                      ^
C:/BW/PlayStation-Build-EWS/build/WebKitBuild/Release/WebCore/PrivateHeaders\WebCore/Document.h:247:7: note: forward declaration of 'WebCore::StringCallback'
  247 | class StringCallback;
      |       ^
2 errors generated.
```

However, it is very odd forward declaring StringCallback does not work here since the class constructor and destructor are both out of line. The data member should not require knowing about the type, even when using RefPtr.

--- a/LayoutTests/platform/wpe/TestExpectations
+++ b/LayoutTests/platform/wpe/TestExpectations
@@ -1370,6 +1370,9 @@ imported/w3c/web-platform-tests/css/css-pseudo/highlight-custom-properties-dynam
 imported/w3c/web-platform-tests/css/css-view-transitions/start-skip-start.html [ Failure ]
 imported/w3c/web-platform-tests/css/css-view-transitions/start-view-transtion-skips-active.html [ Failure ]
 
+# Not supported on this platform.
+security/contentSecurityPolicy/plugins-types-allows-youtube-plugin-replacement.html [ Skip ]
+
 # These tests are passing in WebKitGTK, but failing on WPE (timeout).
 http/tests/websocket/tests/hybi/inspector/before-load.html [ Skip ]
 http/tests/websocket/tests/hybi/inspector/binary.html [ Skip ]

--- a/LayoutTests/security/contentSecurityPolicy/plugins-types-allows-youtube-plugin-replacement.html
+++ b/LayoutTests/security/contentSecurityPolicy/plugins-types-allows-youtube-plugin-replacement.html
@@ -8,11 +8,12 @@ if (window.testRunner) {
     testRunner.waitUntilDone();
 }
 
-// Waiting at least 100ms seems to ensure that YouTube plugin replacement has loaded.
-window.setTimeout(function () {
-    if (window.testRunner)
-        testRunner.notifyDone();
-}, 100);
+if (window.internals) {
+    internals.setConsoleMessageListener((message) => {
+        if (message.startsWith("Blocked access to external URL"))
+            testRunner.notifyDone();
+    });
+}
 </script>
 </head>
 <body>

--- a/Source/WebCore/dom/Document.cpp
+++ b/Source/WebCore/dom/Document.cpp
@@ -8566,12 +8566,8 @@ void Document::addConsoleMessage(MessageSource source, MessageLevel level, const
         postTask(AddConsoleMessageTask(source, level, message));
         return;
     }
-
     if (RefPtr page = this->page())
         page->console().addMessage(source, level, message, requestIdentifier, this);
-
-    if (RefPtr consoleMessageListener = m_consoleMessageListener)
-        consoleMessageListener->scheduleCallback(*this, message);
 }
 
 void Document::addMessage(MessageSource source, MessageLevel level, const String& message, const String& sourceURL, unsigned lineNumber, unsigned columnNumber, RefPtr<Inspector::ScriptCallStack>&& callStack, JSC::JSGlobalObject* state, unsigned long requestIdentifier)
@@ -10615,11 +10611,6 @@ void Document::downgradeReferrerToRegistrableDomain()
         m_referrerOverride = makeString(referrerURL.protocol(), "://"_s, domainString, ':', *port, '/');
     else
         m_referrerOverride = makeString(referrerURL.protocol(), "://"_s, domainString, '/');
-}
-
-void Document::setConsoleMessageListener(RefPtr<StringCallback>&& listener)
-{
-    m_consoleMessageListener = listener;
 }
 
 AnimationTimelinesController& Document::ensureTimelinesController()

--- a/Source/WebCore/dom/Document.h
+++ b/Source/WebCore/dom/Document.h
@@ -1802,8 +1802,6 @@ public:
     const Logger& logger() const { return const_cast<Document&>(*this).logger(); }
     WEBCORE_EXPORT static const Logger& sharedLogger();
 
-    WEBCORE_EXPORT void setConsoleMessageListener(RefPtr<StringCallback>&&); // For testing.
-
     void updateAnimationsAndSendEvents();
     void updateStaleScrollTimelines();
     WEBCORE_EXPORT DocumentTimeline& timeline();
@@ -2483,7 +2481,6 @@ private:
 
     const std::unique_ptr<OrientationNotifier> m_orientationNotifier;
     mutable RefPtr<Logger> m_logger;
-    RefPtr<StringCallback> m_consoleMessageListener;
 
     RefPtr<DocumentTimeline> m_timeline;
     const std::unique_ptr<AnimationTimelinesController> m_timelinesController;

--- a/Source/WebCore/page/PageConsoleClient.cpp
+++ b/Source/WebCore/page/PageConsoleClient.cpp
@@ -62,6 +62,7 @@
 #include "Page.h"
 #include "ScriptableDocumentParser.h"
 #include "Settings.h"
+#include "StringCallback.h"
 #include <JavaScriptCore/ConsoleMessage.h>
 #include <JavaScriptCore/JSCInlines.h>
 #include <JavaScriptCore/ScriptArguments.h>
@@ -136,6 +137,11 @@ void PageConsoleClient::logMessageToSystemConsole(const Inspector::ConsoleMessag
     ConsoleClient::printConsoleMessage(consoleMessage.source(), consoleMessage.type(), consoleMessage.level(), consoleMessage.toString(), consoleMessage.url(), consoleMessage.line(), consoleMessage.column());
 }
 
+void PageConsoleClient::setConsoleMessageListener(RefPtr<StringCallback>&& listener)
+{
+    m_consoleMessageListener = listener;
+}
+
 void PageConsoleClient::addMessage(std::unique_ptr<Inspector::ConsoleMessage>&& consoleMessage)
 {
     Ref page = m_page.get();
@@ -154,6 +160,9 @@ void PageConsoleClient::addMessage(std::unique_ptr<Inspector::ConsoleMessage>&& 
             message = consoleMessage->message();
 
         page->chrome().client().addMessageToConsole(consoleMessage->source(), consoleMessage->level(), message, consoleMessage->line(), consoleMessage->column(), consoleMessage->url());
+
+        if (RefPtr consoleMessageListener = m_consoleMessageListener)
+            consoleMessageListener->invoke(message);
 
         if (page->settings().logsPageMessagesToSystemConsoleEnabled() || shouldPrintExceptions()) [[unlikely]]
             logMessageToSystemConsole(*consoleMessage);
@@ -223,6 +232,9 @@ void PageConsoleClient::messageWithTypeAndLevel(MessageType type, MessageLevel l
 
     if (!messageArgumentsVector.isEmpty()) {
         page->chrome().client().addMessageToConsole(MessageSource::ConsoleAPI, level, messageText, lineNumber, columnNumber, url);
+
+        if (RefPtr consoleMessageListener = m_consoleMessageListener)
+            consoleMessageListener->invoke(messageText);
     }
 
     if (page->settings().logsPageMessagesToSystemConsoleEnabled() || PageConsoleClient::shouldPrintExceptions())

--- a/Source/WebCore/page/PageConsoleClient.h
+++ b/Source/WebCore/page/PageConsoleClient.h
@@ -48,6 +48,7 @@ namespace WebCore {
 
 class Document;
 class Page;
+class StringCallback;
 
 class WEBCORE_EXPORT PageConsoleClient final : public JSC::ConsoleClient, public CanMakeCheckedPtr<PageConsoleClient> {
     WTF_MAKE_TZONE_ALLOCATED_EXPORT(PageConsoleClient, WEBCORE_EXPORT);
@@ -56,12 +57,18 @@ public:
     explicit PageConsoleClient(Page&);
     virtual ~PageConsoleClient();
 
+    PageConsoleClient(const PageConsoleClient&) = delete;
+    PageConsoleClient& operator=(const PageConsoleClient&) = delete;
+    PageConsoleClient(PageConsoleClient&&) = delete;
+    PageConsoleClient& operator=(PageConsoleClient&&) = delete;
+
     static bool shouldPrintExceptions();
     static void setShouldPrintExceptions(bool);
 
     static void mute();
     static void unmute();
 
+    void setConsoleMessageListener(RefPtr<StringCallback>&&); // For testing.
     void addMessage(std::unique_ptr<Inspector::ConsoleMessage>&&);
 
     // The following addMessage function are deprecated.
@@ -90,6 +97,7 @@ private:
     Ref<Page> protectedPage() const;
 
     WeakRef<Page> m_page;
+    RefPtr<StringCallback> m_consoleMessageListener;
 };
 
 } // namespace WebCore

--- a/Source/WebCore/testing/Internals.cpp
+++ b/Source/WebCore/testing/Internals.cpp
@@ -586,6 +586,8 @@ void Internals::resetToConsistentState(Page& page)
     page.setDefersLoading(false);
     page.setResourceCachingDisabledByWebInspector(false);
 
+    page.console().setConsoleMessageListener(nullptr);
+
     RefPtr localMainFrame = page.localMainFrame();
     if (!localMainFrame)
         return;
@@ -749,8 +751,6 @@ Internals::Internals(Document& document)
         setAutomaticLinkDetectionEnabled(false);
         setAutomaticTextReplacementEnabled(true);
     }
-
-    setConsoleMessageListener(nullptr);
 
 #if ENABLE(APPLE_PAY)
     auto* frame = document.frame();
@@ -6662,10 +6662,11 @@ void Internals::updateQuotaBasedOnSpaceUsage()
 
 void Internals::setConsoleMessageListener(RefPtr<StringCallback>&& listener)
 {
-    if (!contextDocument())
+    RefPtr page = contextDocument() ? contextDocument()->page() : nullptr;
+    if (!page)
         return;
 
-    contextDocument()->setConsoleMessageListener(WTFMove(listener));
+    page->console().setConsoleMessageListener(WTFMove(listener));
 }
 
 void Internals::setResponseSizeWithPadding(FetchResponse& response, uint64_t size)

